api/app/Console/Commands/SendNotificationsAdHocEmail.php

```diff
@@ -0,0 +1,148 @@
+                                {templateIdFr : The template ID for the French email message}
+                                {--emailAddress=* : The list of contact email addresses to send the notification to}
+                                {--notificationFamily=* : The list of notification families to send the notification to}
+                                {--notifyAllUsers : Send the notification to all users}
+    ';
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Send ad hoc email notifications from given GC Notify templates';
+
+    /**
+     * Execute the console command.
+     *
+     * @return int
+     */
+    public function handle()
+    {
+        $successCount = 0;
+        $failureCount = 0;
+
+        $users = $this->getUsersToSendNotificationTo();
+        $userCount = $users->count();
+
+        if ($this->confirm('Do you wish to send notifications to '.$userCount.' users?')) {
+            $progressBar = $this->output->createProgressBar($userCount);
+            $notification = new AdHocEmail($this->argument('templateIdEn'), $this->argument('templateIdFr'));
+
+            $users->chunk(200, function (Collection $chunkOfUsers) use (&$successCount, &$failureCount, $progressBar, $notification) {
+                /** @var \App\Models\User $user */
+                foreach ($chunkOfUsers as $user) {
+                    try {
+                        $user->notify($notification);
+                        $successCount++;
+                    } catch (\Throwable $e) {
+                        $this->error("Failed to send notification to user $user->id: ".$e->getMessage());
+                        $failureCount++;
+                    } finally {
+                        $progressBar->advance();
+                    }
+                }
+            });
+            $this->newLine();
+
+            $this->info("Success: $successCount Failure: $failureCount");
+            if ($failureCount > 0) {
+                return Command::FAILURE;
+            } else {
+                return Command::SUCCESS;
+            }
+        } else {
+            $this->info('Notification sending cancelled');
+
+            return Command::SUCCESS;
+        }
+    }
+
+    private function getUsersToSendNotificationTo(): Builder
+    {
+        $emailAddresses = $this->option('emailAddress');
+        $notificationFamilies = $this->option('notificationFamily');
+        $notifyAllUsers = $this->option('notifyAllUsers');
+
+        $options = $this->options();
+
+        // How many types of options were used?
+        $optionTypesCount =
+            (count($emailAddresses) > 0 ? 1 : 0) +
+            (count($notificationFamilies) > 0 ? 1 : 0) +
+            ($notifyAllUsers ? 1 : 0);
+
+        if ($optionTypesCount != 1) {
+            throw new \Error('Must filter users using exactly one of the option types');
```

Comment:
This can simply use InvalidArgumentException .. but your choice :)

Reply:
You're right - that's more descriptive.
[switch to InvalidArgumentException](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12325/commits/957001b96edad72e65c931b3421339f822a4f5f1)